File: 1029-vertical-order-traversal-of-a-binary-tree/1029-vertical-order-traversal-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
    using pii = pair<int, int>;

public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        vector<vector<int>> ans;
        if (!root)
            return ans;
        queue<pair<pii, TreeNode*>> q;
        q.push({{0, 0}, root});    // {{vertical_distance, level}, Node};
        map<int, vector<pii>> map; // [dis] -> {level, val}

        while (!q.empty()) {
            int n = q.size();
            auto dis = q.front().first.first;
            auto level = q.front().first.second;
            auto node = q.front().second;
            q.pop();

            map[dis].push_back({level, node->val});

            if (node->left)
                q.push({{dis - 1, level + 1}, node->left});
            if (node->right)
                q.push({{dis + 1, level + 1}, node->right});
        }

        for (auto it = map.begin(); it != map.end(); it++) {

            sort(it->second.begin(), it->second.end(), [&](const pii& a, const pii& b) {
                if (a.first == b.first)
                    return a.second < b.second;
                return a.first < b.first;
            });

            vector<int> temp;
            // sort(it->second.begin(), it->second.end());
            //cout << it->first << " -> ";
            for (auto& pr : it->second) {
                temp.push_back(pr.second);
                //cout << "[" << pr.second << ", " << pr.first << "], ";
            }
            //cout << endl;

            ans.push_back(temp);
        }

        return ans;
    }
};

File: 1029-vertical-order-traversal-of-a-binary-tree/1029-vertical-order-traversal-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    class Pair {
        int[] first;
        TreeNode second;

        Pair(int[] first, TreeNode second) {
            this.first = first;
            this.second = second;
        }
    }

    public List<List<Integer>> verticalTraversal(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null)
            return ans;

        Queue<Pair> q = new ArrayDeque<>();  //
        q.offer(new Pair(new int[] {0, 0}, root));

        Map<Integer, List<int[]>> map = new TreeMap<>();

        while (!q.isEmpty()) {
            int n = q.size();
            int dis = q.peek().first[0];
            int level = q.peek().first[1];
            TreeNode node = q.peek().second;
            q.poll();

            map.computeIfAbsent(dis, k -> new ArrayList<>()).add(new int[] {level, node.val});

            if (node.left != null)
                q.offer(new Pair(new int[] {dis - 1, level + 1}, node.left));
            if (node.right != null)
                q.offer(new Pair(new int[] {dis + 1, level + 1}, node.right));
        }

        for (Map.Entry<Integer, List<int[]>> em : map.entrySet()) {

            List<int[]> dummy = em.getValue();
            List<Integer> temp = new ArrayList<>();

            Collections.sort(dummy, (a, b) -> {
                if (a[0] == b[0])
                    return a[1] - b[1];
                return a[0] - b[0];
            });

            for (int[] ele : dummy)
                temp.add(ele[1]);

            ans.add(temp);
        }

        return ans;
    }
}